### PR TITLE
Decouple Operations on Order Data From The Rest

### PR DESCRIPTION
**Pull request by pavelbrm**
This PR introduces a set of changes aimed at eventually improving the coherence and health of the SKUs service.

## TL;DR

- First steps towards better structure with packages.
- The order and order items functionality has been moved to `storage/repository` from the massive `skus.Postgres` datastore.
- Data model types have been moved to the `model` package, and made backward-compatible (for now) via type aliases.
- A couple of cosmetic and quality of life improvements, such as neat bulk insert for items instead of the manual loop.

Read on to learn more about this.

## Details

The changes can be grouped in the following few categories.

### Structural

At present, the SKUs  service is made of one package where all different moving parts are together. This poses a number a challenges, such as with testing, development, and maintaining clear and strong responsibility boundaries.

Experience has shown that in an average service, ranging from small to fairly large in size, the following can be found:
- The transport layer – where incoming data is received and converted to a runtime representation stripped off of any transport details. These are handlers.
    - For web/API: it's when a request is received, and data is prepared to be worked on.
    - For consumers/workers: it's when a message is fetched from a stream, and data is prepared to be processed.
- The business-logic layer – where the data in its runtime form is interacted with to perform a business process. These are services (in many places also known as controllers).
    - Database transactions are managed here. It's the business-logic that holds the knowledge about what should happen together, and how a failure should be handled.
- The data access layer – where the data in its runtime form is saved to or is fetched into from a data store. These are repositories.
    - Repositories are oblivious to the notion of a database transaction and must not be concerned with it. Otherwise, reasoning about it as well as testing are complicated.
    - Usually, one repository per table.
- The data model – it's the definitions of the runtime representation of the data and their behaviour.
    - A good place to encapsulate some data-only algorithms and transformations.

In a Go service, this translates to the following packages:
- `handler` – for handlers;
- `service` – for business logic;
- `storage/repository` – for repositories (and migrations normally go to `storage/migration`);
- `model` – for data types.

It's important to note that it's desirable, possible and practically achievable (and rewarding) to be in a situation where only the data layer, `model`, is interacted with by all other three layers (meaning that none of the other three packages import each other, but do import `model`).

This PR starts changing the situation by:
- introducing `storage/repository`, and moving operations on orders and order items there;
- introducing `model`, and moving all data types there. Backward-compatibility for now is ensured via type aliasing (which exists to enable transitions like this).

Eventually, the `skus.Postgres` type should be gone. The business logic will manage transactions (open, commit, rollback), and pass it around to the relevant methods of one or more repositories.

### Dependency Injection

One of the Go's distinctive features – the ability to impose requirements where you need them AKA declare interfaces where you use them, exists for a reason.

_This is the way_ to loose coupling, ease of testing, and other pleasant features of a good code base.

For example, a service does not need to know about a concrete repository or some central interface. It needs something that can work with orders. The service declares what it needs. Likewise, a handler does not need to know about a concrete service. All it needs is a set of behaviours. And so on.

Types that depend on other things can choose to accept dependencies via:
- being passed it as an argument in a method; use this for "request scoped" dependencies; examples:
    - the famous `ServeHTTP` with its `ResponseWriter`
    - repositories with a handle to the database (either transaction or client, which both implement same interfaces in the case of `database/sql` or `sqlx`)
- being passed it in a constructor and held as a field; use this for long-lived dependencies and situations where the dependency should not be part of the contract; examples:
    - a database transaction needs a reference to the client, but this fact should be transparent for methods on transactions (see `sql.Tx` or `sqlx.Tx`)
    - an order handler needs one order service for requests concerning orders;
    - an order service needs one order repository for queries concerning orders.

The changes in this PR will result in significant simplification of tests. Additionally, the need to use frameworks for mocking and code generation will be gone.

I deliberately not listed `context.Context` as a way for dependency injection because it is not. It might be used in rare occasions to circumvent existing interfaces (like injecting request id generated by a server), but those should be unpacked ASAP as close as possible to the place of origin and passed around explicitly.

### Quality of Life Improvements.

A few things fall under this category:
1. `sqlx` offers convenient ways for performing queries, including bulk queries through name binding.
2. Pure data types (`model`) can and should use methods to express various steps of business logic.
3. Testing using the real database can and should be limited to the code inside the `storage/repository` package. Once it's covered, code that depends on those, uses mocks for testing.
4. Table-driven tests are great, with test cases declared as a set of given inputs and expected outputs.
5. Subtests are great for running test cases.

### Type Of Change

- [ ] Product feature
- [ ] Bug fix
- [ ] Performance improvement
- [x] Refactor
- [ ] Other

### Tested Environments

- [ ] Development
- [ ] Staging
- [ ] Production

### Before submitting this PR:

- [x] Does your code build cleanly without any errors or warnings?
- [ ] Have you used auto closing keywords?
- [x] Have you added tests for new functionality?
- [ ] Have validated query efficiency for new database queries?
- [ ] Have documented new functionality in README or in comments?
- [x] Have you squashed all intermediate commits?
- [x] Is there a clear title that explains what the PR does?
- [x] Have you used intuitive function, variable and other naming?
- [ ] Have you requested security / privacy review if needed
- [x] Have you performed a self review of this PR?

---

### CODE REVIEW

**Review by husobee — DISMISSED** [review]
looking forward to implementing this structure in our other services in bat-go.

**Review by husobee — DISMISSED** [review]

File: services/skus/storage/repository/repository.go [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+	"github.com/shopspring/decimal"
+
+	"github.com/brave-intl/bat-go/libs/datastore"
+
+	"github.com/brave-intl/bat-go/services/skus/model"
+)
+
+type Order struct{}
+
+func NewOrder() *Order { return &Order{} }
```
Comment (clD11) [inline services/skus/storage/repository/repository.go:22]: Maybe I am missing something/interpreting this wrong, but where a type only exists to implement an interface i.e. will never have exported methods beyond the interface do we not want to return the interface. For example, this constructor would return OrderStore interface as there is no need to export the type itself. This signals there is nothing of interest beyond the interface.

File: services/skus/storage/repository/order_history.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+	"github.com/jmoiron/sqlx"
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/brave-intl/bat-go/services/skus/model"
+)
+
+type OrderPayHistory struct{}
+
+func NewOrderPayHistory() *OrderPayHistory { return &OrderPayHistory{} }
```
Comment (clD11) [inline services/skus/storage/repository/order_history.go:15]: same questions about interfaces

File: services/skus/storage/repository/order_item.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+	"github.com/jmoiron/sqlx"
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/brave-intl/bat-go/services/skus/model"
+)
+
+type OrderItem struct{}
+
+func NewOrderItem() *OrderItem { return &OrderItem{} }
```
Comment (clD11) [inline services/skus/storage/repository/order_item.go:16]: same questions about interfaces

File: services/skus/storage/repository/repository_test.go [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+	}
+
+	type testCase struct {
+		name  string
+		given int64
+		exp   tcExpected
+	}
+
+	tests := []testCase{
+		testCase{
```
[the remark this answers is not in the document]
Comment (pavelbrm) [reply services/skus/storage/repository/repository_test.go]: Resolving this.

File: services/skus/storage/repository/order_item_test.go [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+					Subtotal:       mustDecimalFromString("2"),
+					CredentialType: "something",
+				},
+			},
+		},
+
+		testCase{
+			name: "two_items",
+			given: []model.OrderItem{
+				model.OrderItem{
```
[the remark this answers is not in the document]
Comment (pavelbrm) [reply services/skus/storage/repository/order_item_test.go]: Addressed. Resolving this.

File: services/skus/storage/repository/order_item_test.go [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+					Quantity:       1,
+					Price:          mustDecimalFromString("2"),
+					Currency:       "USD",
+					Subtotal:       mustDecimalFromString("2"),
+					CredentialType: "something",
+				},
+			},
+
+			exp: []model.OrderItem{
+				model.OrderItem{
```
[the remark this answers is not in the document]
Comment (pavelbrm) [reply services/skus/storage/repository/order_item_test.go]: Addressed. Resolving this.

File: services/skus/storage/repository/repository.go (hunk 1 above)
Comment (pavelbrm) [reply services/skus/storage/repository/repository.go:22]: Just a note that we've discussed this, and decided to keep as it is.

**Review by clD11 — APPROVED** [review]
